pkg/serverless/frequency.go

```diff
@@ -0,0 +1,73 @@
+package serverless
+
+import (
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/serverless/flush"
+)
+
+const maxInvocationsStored = 50
+
+// StoreInvocationTime stores the given invocation time in the list of previous
+// invocations. It is used to compute the invocation frequency of the current function.
+// It is automatically removing entries when too much have been already stored (more than maxInvocationsStored).
+// When trying to store a new point, if it is older than the last one stored, it is ignored.
+// Returns if the point has been stored.
+func (d *Daemon) StoreInvocationTime(t time.Time) bool {
+	// ignore points less recent than the last stored one
+	if len(d.lastInvocations) > 0 && d.lastInvocations[len(d.lastInvocations)-1].After(t) {
+		return false
+	}
+
+	// remove when too much/old entries
+	d.lastInvocations = append(d.lastInvocations, t)
+	if len(d.lastInvocations) > maxInvocationsStored {
+		d.lastInvocations = d.lastInvocations[len(d.lastInvocations)-maxInvocationsStored : len(d.lastInvocations)]
+	}
+
+	return true
+}
+
+// InvocationFrequency computes the invocation frequency of the current function.
+// This function returns 0 if not enough invocations were done.
+func (d *Daemon) InvocationFrequency() time.Duration {
+	// with less than 3 invocations, we don't have enough data to compute
+	// something reliable.
+	if len(d.lastInvocations) < 3 {
+		return 0
+	}
+
+	var total int64
+	for i := 1; i < len(d.lastInvocations); i++ {
+		total += int64(d.lastInvocations[i].Sub(d.lastInvocations[i-1]))
+	}
+
+	return time.Duration(total / int64(len(d.lastInvocations)-1))
+}
+
+// AutoSelectStrategy uses the invocation frequency of the function to select the
+// best flush strategy.
+// This function doesn't mind if the flush strategy has been overridden through
+// configuration / environment var, the caller is responsible of that.
+func (d *Daemon) AutoSelectStrategy() flush.Strategy {
+	freq := d.InvocationFrequency()
```

Comment:
Seconding Darcy's comment, I think this is more of an interval (length of time) rather than a frequency (rate per amount of time)

Reply:
Correct.